### PR TITLE
Implement `safe_glob` that ignores "." and ".."

### PR DESCRIPTION
**Pull request by keszybz**
This is untested, except for units tests, so please don't use this on #5644 just yet.
I'm putting this out here in case anyone wants to comment on the general direction.

### CODE REVIEW

**Comment by poettering** [comment]
I like it. Thinking about this I'd proably call it safe_glob() rather than glob_safe() though, following our usual scheme of safe_close(), safe_atou32() and so on.

Also, this call manipulates the glob_t passed in, in a way that isn't expected I figure. Hence I think it would be good to add a big comment about that near the safe_glob() implementation or so.

Hmm, and if you want to be entirely accurate, then it might make sense to override the function fields not just when they are NULL, but also when GLOB_ALTDIRFUNC isn't set. Because otherwise if a caller sets any of the functions, and then doesn't pass the flag we silently turn the function on even if it originally wouldn't be called... But I figure that particular case isn't too crucial

**Comment by keszybz** [comment]
Updated, with the following changes:
- glob_safe is renamed to safe_glob
- I added comments and an assert that GL_ALTDIRFUNC is not set in flags.

It seems to fix #5644 quite nicely:
```console
$ cat /etc/tmpfiles.d/test.conf
R /foo/.* - - - - -
R /
R /../
R /../././
$ sudo SYSTEMD_LOG_LEVEL=debug systemd-tmpfiles --remove /etc/tmpfiles.d/test.conf
Successfully loaded SELinux database in 4.792ms, size on heap is 909K.
Reading config file "/etc/tmpfiles.d/test.conf".
Running remove action for entry R /foo/.*
rm -rf "/foo/.bar1"
rm -rf "/foo/.bar2"
Running remove action for entry R /
rm -rf "/"
Attempted to remove entire root file system, and we can't allow that.
rm_rf(/): Operation not permitted
Running remove action for entry R /..
rm -rf "/.."
Attempted to remove entire root file system, and we can't allow that.
rm_rf(/..): Operation not permitted
Running remove action for entry R /.././.
rm -rf "/.././."
Attempted to remove entire root file system, and we can't allow that.
rm_rf(/.././.): Operation not permitted
```

So I think this can go in, CI gods permitting.

**Comment by yuwata** [comment]
Still glob_safe instead of safe_glob is used in the commit comments.

**Comment by keszybz** [comment]
On Wed, Apr 26, 2017 at 09:02:46PM -0700, Yu Watanabe wrote:
> Still glob_safe instead of safe_glob is used in the commit comments.
Thanks, updated.

**Comment by poettering** [comment]
lgtm. CI doesn't like it though, but I figure that's not actually this PR's fault, but something in the meson context caused it.

**Comment by martinpitt** [comment]
I retried the s390x build after the libudev linking fix, and it's green now. Semaphore failed only on the meson build - that's fixed in master, but semaphore doesn't rebase PRs against master, so it's expected.

File: src/core/execute.c [hunk 1]
```diff
@@ -3253,23 +3252,19 @@ int exec_context_load_environment(Unit *unit, const ExecContext *c, char ***l) {
...
-                                continue;
 
-                        strv_free(r);
-                        return -EINVAL;
-                }
-                for (n = 0; n < count; n++) {
+                /* When we don't match anything, -ENOENT should be returned */
+                assert(pglob.gl_pathc > 0);
+
+                for (n = 0; n < pglob.gl_pathc; n++) {
```
Comment (evverx) [inline src/core/execute.c:3267]: This causes the following warning
```
../src/core/execute.c: In function ‘exec_context_load_environment’:
../src/core/execute.c:3267:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                 for (n = 0; n < pglob.gl_pathc; n++) {
                               ^
```
.

File: src/basic/rm-rf.c [hunk 2]
```diff
@@ -182,18 +182,11 @@ int rm_rf(const char *path, RemoveFlags flags) {
         /* We refuse to clean the root file system with this
          * call. This is extra paranoia to never cause a really
          * seriously broken system. */
-        if (path_equal(path, "/")) {
+        if (path_equal_or_files_same(path, "/")) {
```
Comment (evverx) [inline src/basic/rm-rf.c:185]: `files_same` follows symlinks, so it is not possible to remove a symlink to `/` with `rm_rf`.

Comment (poettering) [reply src/basic/rm-rf.c:185]: I figure we should a flavour of files_same() that doesn't follow symlinks, and similar a flavour of path_equal_or_files_same(), i.e. honouring a flag AT_SYMLINK_NOFOLLOW or so

Comment (poettering) [reply src/basic/rm-rf.c:185]: I filed #5864 now about this.

**Pull request merged**